### DISCUSSION TITLE
How Windhawk can be deployed with pre-set mods?

### DISCUSSION THREAD

**Discussion by Anixx** [ramensoftware/windhawk#155] [General]
Can Windhawk be quickly deployed to a new system witha set of pre-configured mods?

**Comment by m417z** [comment]
I was thinking about adding an export/import feature, but currently, there's no official way. You can install the mods manually once, and then use that Windhawk instance for the deploy. For the non-portable version, both files from the program data folder and registry entries will have to be copied.
I haven't tried it, but I believe that it's enough to copy over the following:
```
%ProgramData%\Windhawk\Engine\Mods
%ProgramData%\Windhawk\ModsSource
HKLM\SOFTWARE\Windhawk\Engine\Mods
HKLM\SOFTWARE\Windhawk\Engine\ModsWritable
```

Let me know whether that works for you.

**Reply by DepressedOptimist** [reply]
Just wanted to confirm that this method works.

Note that I didn't have a ModsSource folder. Also, I exported the entire HKEY_LOCAL_MACHINE\SOFTWARE\Windhawk because I noticed that the Process Exclusion List was under HKEY_LOCAL_MACHINE\SOFTWARE\Windhawk\Settings.

Not issues so far.

**Reply by OrthodoxWindows** [reply]
I'm having a issue with this method. The symbols aren't downloading, which prevents mods that require symbols from working.
Mods that don't depend on symbols work without any issue.

**Reply by m417z** [reply]
@OrthodoxWindows I can't think of why it should affect symbol downloading. Can you post logs for one of the mods using the instruction on [the troubleshooting page](https://github.com/ramensoftware/windhawk/wiki/Troubleshooting#:~:text=Pick%20a%20mod,to%20None.)?

**Reply by Chaython** [reply]
> I was thinking about adding an export/import feature, but currently, there's no official way. You can install the mods manually once, and then use that Windhawk instance for the deploy. For the non-portable version, both files from the program data folder and registry entries will have to be copied. I haven't tried it, but I believe that it's enough to copy over the following:
> 
> ```
> %ProgramData%\Windhawk\Engine\Mods
> %ProgramData%\Windhawk\ModsSource
> HKLM\SOFTWARE\Windhawk\Engine\Mods
> HKLM\SOFTWARE\Windhawk\Engine\ModsWritable
> ```
> 
> Let me know whether that works for you.

Please add an import/export or cloud sync. [Uniget added a cloudsync via github]

**Reply by rickhantz-orig** [reply]
Backup works, haven't tried restore yet.
Any problem backing up whole WindHawk reg key?

BackupWindHawk.bat:
xcopy %ProgramData%\Windhawk\Engine\Mods .\Engine\Mods\
xcopy %ProgramData%\Windhawk\ModsSource .\ModsSource\
reg export HKLM\SOFTWARE\Windhawk .\Reg\WindhawkHKLM.reg /y
pause

RestoreWindHawk.bat:
rem @echo off
IF EXIST ".\Windhawk\Engine\Mods" (
   xcopy  .\Windhawk\Engine\Mods\ %ProgramData%\Windhawk\Engine\Mods\
) ELSE (
    ECHO Run backup first/no files found!
)

IF EXIST "\Windhawk\ModsSource" (
   xcopy  .\Windhawk\ModsSource\ %ProgramData%\Windhawk\ModsSource\
) ELSE (
    ECHO Run backup first/no files found!

reg import .\Reg\WindhawkHKLM.reg
pause

**Reply by jcbyte** [reply]
Based on @rickhantz-orig comment.
Backup and Restore works perfectly for me in my unattended setup script with these PowerShell scripts:

Save mods and registry entries into an archive:
```ps
$ArchiveFile = "windhawk-config-archive.zip"

$TempDir = Join-Path $env:TEMP ([guid]::NewGuid())
New-Item -ItemType Directory -Path $TempDir | Out-Null

$WindhawkDir = "$env:PROGRAMDATA\Windhawk"
$CollectedModsDir = Join-Path $TempDir "Mods"
Copy-Item "$WindhawkDir\Engine\Mods" -Destination "$CollectedModsDir\Engine\Mods" -Recurse -Force
Copy-Item "$WindhawkDir\ModsSource" -Destination "$CollectedModsDir\ModsSource" -Recurse -Force

$CollectedRegDir = Join-Path $TempDir "Reg"
New-Item -Path $CollectedRegDir -ItemType Directory -Force | Out-Null
reg export "HKLM\SOFTWARE\Windhawk\Engine\Mods" "$CollectedRegDir\Engine-Mods.reg" /y | Out-Null
reg export "HKLM\SOFTWARE\Windhawk\Engine\ModsWritable" "$CollectedRegDir\ENgine-ModsWritable.reg" /y | Out-Null

$Items = Get-ChildItem -Path $TempDir -Force | ForEach-Object { $_.FullName }
Compress-Archive -Path $Items -DestinationPath $ArchiveFile -Force
```

And restore from this archive:
```ps
# Expand the archive into a temporary folder
$WindhawkArchive = "windhawk-config-archive.zip" # Your created backup archive

$TempDir = Join-Path $env:TEMP ([guid]::NewGuid())
New-Item -ItemType Directory -Path $TempDir | Out-Null

Expand-Archive -Path $WindhawkArchive -DestinationPath $TempDir 

# Copy the mods into the windhawk folder
$WindhawkDir = "$env:PROGRAMDATA\Windhawk"
$CollectedModsDir = Join-Path $TempDir "Mods"
Copy-Item "$CollectedModsDir\Engine\Mods" -Destination "$WindhawkDir\Engine\Mods" -Recurse -Force
Copy-Item "$CollectedModsDir\ModsSource" -Destination "$WindhawkDir\ModsSource" -Recurse -Force

# Copy the registry back
$CollectedRegDir = Join-Path $TempDir "Reg"
reg import "$CollectedRegDir\Engine-Mods.reg" | Out-Null
reg import "$CollectedRegDir\ENgine-ModsWritable.reg" | Out-Null

# Cleanup the created temporary folder
Remove-Item $TempDir -Recurse -Force -ErrorAction SilentlyContinue
```

**Reply by sangammagariaravind123** [reply]
That helped share mods but didn't copy the customisations done in yaml file of each mod. (The settings page of each mod where we can edit the yaml file to customize the parameters - control styles, res vars and style consts). Or did I do some mistake 🤔.

**Reply by jcbyte** [reply]
@sangammagariaravind123 Customisations to mods are saved in the registry `HKLM\SOFTWARE\Windhawk\Engine\Mods`, so they should be copied over with this script. Admitted none of my mods have modified settings so I didn't test this.

**Comment by e-t-l** [comment]
This might be a different question entirely, but I wonder if it would be possible to "export" mods as executables, kind of how AutoHotKey can export AHK scripts as executables that can be run without needing a AutoHotKey.exe installed separately on the system already. I'm imagining if there were a way to analyze a mod and figure out which WH components/hooks/etc are utilized, then package the mod with just those basic WH components/framework required to make it work as a standalone executable...

**Reply by m417z** [reply]
It can be possible, but I don't have plans to implement it. Instead, I have it on my list to have a "lite" version of Windhawk without VSCode and a compiler, so that its size would be around 10 MB instead of around 700 MB. It won't allow modifying or developing mods, but it will be easier and quicker to install for most users. Perhaps it will cover your needs as well.

**Reply by stamminator** [reply]
@m417z I'm working on a guide for "I want a better Windows experience but I'm not really a power user and don't want to tinker" users, and that kind of Windhawk lite version would be perfectly suited for it.

**Reply by happymimimix** [reply]
> This might be a different question entirely, but I wonder if it would be possible to "export" mods as executables, kind of how AutoHotKey can export AHK scripts as executables that can be run without needing a AutoHotKey.exe installed separately on the system already. I'm imagining if there were a way to analyze a mod and figure out which WH components/hooks/etc are utilized, then package the mod with just those basic WH components/framework required to make it work as a standalone executable...

I also want to export mods as a standalone executable that could work without windhawk!

**Reply by decipher3114** [reply]
> It can be possible, but I don't have plans to implement it. Instead, I have it on my list to have a "lite" version of Windhawk without VSCode and a compiler, so that its size would be around 10 MB instead of around 700 MB. It won't allow modifying or developing mods, but it will be easier and quicker to install for most users. Perhaps it will cover your needs as well.

Instead of packing the whole VSCodium, why not include a simple server in the windhawk itself, which can be commnunicated by the `windhawk.net` website.

It will be much lighter and faster.

**Reply by stamminator** [reply]
@decipher3114 There's nothing wrong with the Windhawk GUI. Bundling a stripped down VS Codium makes sense since the GUI involves things like viewing source code, diffs, and editing config files.

**Reply by Anixx** [reply]
Diffs? Where are the diffs there?

**Reply by decipher3114** [reply]
@stamminator `windhawk.net` does the same thing. (and it's more responsive)
